server/src/main/java/com/flipkart/varadhi/authn/CustomAuthenticationHandler.java

```diff
@@ -0,0 +1,171 @@
+                );
+            }
+            authenticationProvider = (AuthenticationProvider)providerClass.getDeclaredConstructor().newInstance();
+
+            try {
+                authenticationProvider.init(authenticationOptions, orgResolver, meterRegistry);
+            } catch (Exception e) {
+                throw new InvalidConfigException("Failed to initialize authenticator: " + e.getMessage(), e);
+            }
+
+        } catch (ClassNotFoundException e) {
+            throw new InvalidConfigException(
+                "Authentication provider class not found: " + authenticationOptions
+                                                                                   .getAuthenticationProviderClassName(),
+                e
+            );
+        } catch (ReflectiveOperationException e) {
+            throw new InvalidConfigException("Failed to create authentication provider", e);
+        }
+
+        return new CustomAuthenticationHandler(
+            authenticationProvider,
+            new URLMatcherUtil(authenticationOptions.getOrgContextExemptionURLs())
+        );
+    }
+
+    @Override
+    public void handle(RoutingContext routingContext) {
+
+        RequestContext requestContext;
+        try {
+            requestContext = createRequestContext(routingContext);
+        } catch (URISyntaxException e) {
+            throw new BadRequestException(e);
+        }
+
+        String orgName = readOrgNameFromContext(routingContext);
+
+        Future<UserContext> userContext = authenticationProvider.authenticate(orgName, requestContext);
+
+        userContext.onComplete(result -> {
+            if (result.succeeded()) {
+                routingContext.put(USER_CONTEXT, result.result());
+                routingContext.next();
+            } else {
+                routingContext.fail(UNAUTHORIZED.code(), result.cause());
+            }
+        });
+    }
+
+    private String readOrgNameFromContext(RoutingContext routingContext) {
+
+        Map<ResourceType, ResourceHierarchy> typeHierarchyMap = routingContext.get(CONTEXT_KEY_RESOURCE_HIERARCHY);
+        if (typeHierarchyMap != null) {
+            if (typeHierarchyMap.containsKey(ResourceType.ORG)) {
+                ResourceHierarchy hierarchy = typeHierarchyMap.get(ResourceType.ORG);
+                if (hierarchy != null && hierarchy.getAttributes() != null && hierarchy.getAttributes()
+                                                                                       .containsKey(TAG_ORG)) {
+                    return hierarchy.getAttributes().get(TAG_ORG);
+                }
+            }
+        }
+
+        if (!orgExemptionURLMatcher.matches(
+            String.valueOf(routingContext.request().method()),
+            routingContext.request().path()
+        )) {
+            throw new UnAuthenticatedException("Org context missing in the request");
```

Comment:
is it unauthenticated or internalServerError?

basically, is it client's issue or the server's issue that the org context is missing?

Reply:
Right. should be a server issue in this case. updated the error.